docs/writing_tests.md

```diff
@@ -0,0 +1,220 @@
+need to add support for the system call you wish to test to linux-api.
+
+These tests are run emulated within Shadow and natively outside of Shadow. This
+is done using the CMake `add_linux_tests` and `add_shadow_tests` macros. The
+tests are built by Cargo and then run by CMake. For example the `listen` tests
+use:
+
+```cmake
+add_linux_tests(BASENAME listen COMMAND sh -c "../../../target/debug/test_listen --libc-passing")
+add_shadow_tests(BASENAME listen)
+```
+
+which results in the CMake tests:
+
+```text
+1/2 Test #110: listen-shadow ....................   Passed    0.56 sec
+2/2 Test #109: listen-linux .....................   Passed   10.12 sec
+```
+
+[libc]: https://docs.rs/libc/latest/libc/
+[linux-api]: https://shadow.github.io/docs/rust/linux_api/
+[domain-tests]: writing_tests.md#domain-specific-system-call-tests
+[fuzz-tests]: writing_tests.md#fuzz-tests
+
+### Domain-specific system call tests
+
+Here is an example of an existing test for the
+[`listen`](https://man7.org/linux/man-pages/man2/listen.2.html) system call:
+
+```rust
+/// Test listen using a backlog of 0.
+fn test_zero_backlog(
+    domain: libc::c_int,
+    sock_type: libc::c_int,
+    flag: libc::c_int,
+    bind: Option<SockAddr>,
+) -> Result<(), String> {
+    let fd = unsafe { libc::socket(domain, sock_type | flag, 0) };
+    assert!(fd >= 0);
+
+    if let Some(address) = bind {
+        bind_fd(fd, address);
+    }
+
+    let args = ListenArguments { fd, backlog: 0 };
+
+    let expected_errno = match (domain, sock_type, bind) {
+        (libc::AF_INET, libc::SOCK_STREAM, _) => None,
+        (libc::AF_UNIX, libc::SOCK_STREAM | libc::SOCK_SEQPACKET, Some(_)) => None,
+        (libc::AF_UNIX, libc::SOCK_STREAM | libc::SOCK_SEQPACKET, None) => Some(libc::EINVAL),
+        (_, libc::SOCK_DGRAM, _) => Some(libc::EOPNOTSUPP),
+        _ => unimplemented!(),
+    };
+
+    test_utils::run_and_close_fds(&[fd], || check_listen_call(&args, expected_errno))
+}
+```
+
+There are many `listen` tests including the one above, such as
+`test_zero_backlog`, `test_negative_backlog`, `test_large_backlog`,
+`test_listen_twice`, `test_reduced_backlog`, and more.
+
+### Fuzz tests
+
+"Fuzz"-style testing is another way we test syscalls: they use some [support
+code][fuzz-support] to test many various combinations of the possible argument
+values expected by a syscall, and verify that the return value for each
+combination of arguments is the same as what Linux returns. Because the
+developer usually writes these tests to cover most or all possible argument
+combinations, it ensures that Shadow's emulation of the syscall is highly
+accurate.
+
+Fuzz tests can be a bit trickier to write, especially for more complicated
+syscalls, and sometimes they don't make sense (e.g., when testing what happens
+when trying to `connect()` to a TCP server with a full accept queue). Thus, we
+don't explicitly require them. However, they often help us find inconsistent
```

Comment:
Is this still consistent with our summary list from `contributing.md`?

> System call tests (domain-specific and fuzz tests) (**required**)

Is this maybe better?:
```diff
- queue). Thus, we don't explicitly require them. However, they
+ queue). They
```

Reply:
Okay yeah I agree that's better.